### PR TITLE
Fix ctr exec stuck problem

### PR DESCRIPTION
**Pull request by Tim-Zhang**
agent: 
- Close stdin when write_stdin receives data of length 0.
- Stop call notify_term_close() in close_stdin, because it could discard stdout unexpectedly.

runtime-rs:
- Instead of call agent.close_stdin in close_io, we call agent.write_stdin with 0 len data when the stdin pipe ends.

### CODE REVIEW

File: src/agent/rustjail/src/process.rs [hunk 1]
```diff
@@ -210,17 +210,8 @@ impl Process {
...
 
     pub async fn close_stdin(&mut self) {
-        if let Some(proc_io) = &mut self.proc_io {
-            // notify io copy task to close stdin stream
-            let _ = proc_io.close_stdin_tx.send(true);
-            // wait for io copy task to finish
-            proc_io.wg_input.wait().await;
-        }
-
         close_process_stream!(self, term_master, TermMaster);
```
[the remark this answers is not in the document]
Comment (Tim-Zhang) [reply src/agent/rustjail/src/process.rs:213]: Yes, they could be closed. In passfd-io, we don't keep term_master and parent_stdin in Process struct so close_process_stream! is meaningless for passfd-io.

File: src/runtime-rs/crates/runtimes/virt_container/src/container_manager/io/container_io.rs [hunk 2]
```diff
@@ -80,6 +83,29 @@ impl<'inner> ContainerIoWrite<'inner> {
             }
         }
     }
+
+    fn poll_shutdown_inner(&'inner mut self, cx: &mut Context<'_>) -> Poll<io::Result<()>> {
+        let mut shutdown_future = self.shutdown_future.take();
+        if shutdown_future.is_none() {
```
[the remark this answers is not in the document]
Comment (Tim-Zhang) [reply src/runtime-rs/crates/runtimes/virt_container/src/container_manager/io/container_io.rs:90]: done

**Review by sprt — COMMENTED** [review]
Related to #9187. This is a tricky code path as we keep regressing here. We should have integration tests to address #1160 and #9186 to have more confidence in our changes.

**Comment by Tim-Zhang** [comment]
> Related to #9187. This is a tricky code path as we keep regressing here. We should have integration tests to address #1160 and #9186 to have more confidence in our changes.

@sprt I have added a stdio CI in this pr, but unfortunately I can't see it in the gh list, maybe it only can be seen after the merge?

**Comment by sprt** [comment]
@Tim-Zhang We already have the infra to support such tests, you don't have to modify any workflow YAML 🙂 I recommend simply adding your test cases to this file: https://github.com/kata-containers/kata-containers/blob/main/tests/integration/kubernetes/k8s-exec.bats Please do reach out on Slack if you need any more guidance here, I'm happy to help.

**Comment by Tim-Zhang** [comment]
@sprt Since the stdio CI tests is aimed to test `ctr` not `kubectl`, I'm not sure if it's appropriate to put it in the k8s-exec.bats file.

**Comment by sprt** [comment]
@Tim-Zhang I would reimplement your tests to use `kubectl` instead of `ctr` as those both repro the issues in the same way in my experience - referring to e.g. #1160 which repros using `kubectl exec`.

**Comment by Tim-Zhang** [comment]
@sprt There are subtle differences between `ctr` and `kubectl`, and `kubectl exec` would not stuck on this issue.

Now that you want to reimplement the tests later how about letting this PR in first?

File: src/agent/src/rpc.rs [hunk 3]
```diff
@@ -584,24 +584,31 @@ impl AgentService {
         let cid = req.container_id;
         let eid = req.exec_id;
 
-        let writer = {
+        let mut resp = WriteStreamResponse::new();
+        resp.set_len(req.data.len() as u32);
+
+        // EOF of stdin
+        if req.data.len() == 0 {
```
Comment (Apokleos) [inline src/agent/src/rpc.rs:591]: Some CIs report error that `req.data.len()== 0` should be changed to `req.is_empty()` to make CIs happy.

Comment (Apokleos) [reply src/agent/src/rpc.rs:591]: sorry, req.is_empty() -> req.data.is_empty()

Comment (Tim-Zhang) [reply src/agent/src/rpc.rs]: done, thanks

**Review by Apokleos — APPROVED** [review]
Thx Tim @Tim-Zhang LGTM

**Review by lifupan — APPROVED** [review]

**Comment by stevenhorsman** [comment]
Hi @Tim-Zhang, I'm a bit confused by this PR, it's trying to do:
```
function run() {
	info "Running stdio tests"

	bats "${runk_dir}/stdio-tests.bats"
}
```
but we don't have any stdio-tests.bats file, so this is failing every time. Was there supposed to be a follow up to this, or should we revert the tests part of this PR?

**Pull request merged**